File: Python/Leetcode Medium/102BinaryTreeLevelOrderTraversal.py
#https://leetcode.com/problems/binary-tree-level-order-traversal/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ret = []
        queue = deque([(root, 0)])
        while len(queue) > 0:
            node, level = queue.popleft()
            if node == None: continue
            if len(ret) == level: ret.append([node.val])
            else: ret[level].append(node.val)
            queue.append((node.left, level+1))
            queue.append((node.right, level+1))
        return ret

File: C++/Leetcode Medium/102BinaryTreeLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <list>
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        vector<vector<int>> res; 
        list<TreeNode*> queue = {root};
        while (!queue.empty()) {
            vector<int> layer;
            int n = queue.size();
            for (int i = 0; i < n; ++i) {
                TreeNode* curr = queue.front();
                queue.pop_front();
                layer.push_back(curr->val);
                if (curr->left) queue.push_back(curr->left);
                if (curr->right) queue.push_back(curr->right);
            }
            res.push_back(layer);
        }
        return res;
    }
};